### DISCUSSION TITLE
useQuery with side effects

### DISCUSSION THREAD

**Discussion by seanaye** [urql-graphql/urql#2491] [Q&A] [answered]
I am using urql in an react + electron app which requires some imperative side effects with the result of a query. Doing these side effects is a little expensive so its best to only do when the actual true equality of the data changes. This is my current solution (roughly). It's not very elegant in my opinion so im wondering if there is a better, recommended approach. 

```ts
const context = { requestPolicy: "cache-and-network" } as const
function MyComponent() {
  const [result] = useQuery(QUERY_DOC, MY_VARS, context)

  // crude memoization
  const stringified = useMemo(() => JSON.stringify(result.data?.someProperty ?? null), [result])
  const memoed = useMemo(() => JSON.parse(stringified), [stringified])

  useEffect(() => {
    if (memoed) {
      sendToMainProcess(memoed)
    }
  }, [memoed])
}
```

**Comment by JoviDeCroock** [comment, marked as answer]
Both caches should give you ref equality, alternatively you can use result.operation.key which is what we use internally to identify unique combination of query and variables. Worth mentioning that key does not change when a mutation, ... updates said query

But generally both norm and doc cache should get you t his consistency in ref equality

**Reply by seanaye** [reply]
I'm not sure I understand the last part. Do you mean that `useEffect(() => {}, [result.data?.some?.deeply?.nested?.property])` will maintain ref equality even though the root `result` object is changing? I am using `@urql/exchange-graphcache` if that helps.

**Reply by JoviDeCroock** [reply]
It should yes as long as no mutation is changing properties in that data the data should stay ref equal
